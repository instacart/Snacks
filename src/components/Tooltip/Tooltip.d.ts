import * as React from 'react'

export interface TooltipProps {
  children?: React.ReactNode

  size?: 'small' | 'medium' | 'large'

  placement?: 'top' | 'left' | 'right' | 'bottom'

  style?: {
    border?: string
    padding?: string
    boxShadow?: string
  }

  arrowStyle?: {
    border?: string
    boxShadowRight?: string
    boxShadowBottom?: string
    boxShadowLeft?: string
    boxShadowTop?: string
  }

  target: React.ReactNode

  snacksStyle: 'primary' | 'secondary' | 'dark'

  onDismiss(): void

  onShow(): void

  isVisible?: boolean
}

declare const Tooltip: React.ComponentClass<TooltipProps>

export default Tooltip
