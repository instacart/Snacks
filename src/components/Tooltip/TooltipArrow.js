import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import colors from '../../styles/colors'

const styles = {
  arrow: {
    position: 'absolute',
    display: 'block',
    height: '12px',
    width: '12px',
  },
}

const RESOLVED_STYLE = {
  primary: {
    backgroundColor: colors.GREEN_500,
  },
  secondary: {
    backgroundColor: '#FFF',
  },
  dark: {
    backgroundColor: colors.GRAY_20,
  },
}

const RESOLVED_PLACEMENT = {
  top: { transform: 'rotate(45deg)' },
  bottom: { transform: 'rotate(45deg)' },
  right: { transform: 'rotate(-45deg)' },
  left: { transform: 'rotate(-45deg)' },
}

const RESOLVE_BORDER_COLOR = {
  primary: colors.GREEN_500,
  secondary: colors.GRAY_74,
  dark: colors.GRAY_20,
}

const resolveStylePlacementBorders = (style, arrowStyle, placement) => {
  const borderColor = RESOLVE_BORDER_COLOR[style]
  const borderStyle =
    arrowStyle && arrowStyle.border ? arrowStyle.border : `1px solid ${borderColor}`

  const boxShadowRight = arrowStyle && arrowStyle.boxShadowRight
  const boxShadowBottom = arrowStyle && arrowStyle.boxShadowBottom
  const boxShadowLeft = arrowStyle && arrowStyle.boxShadowLeft
  const boxShadowTop = arrowStyle && arrowStyle.boxShadowTop

  switch (placement) {
    case 'top':
      return {
        borderRight: borderStyle,
        borderBottom: borderStyle,
        boxShadow: boxShadowTop,
      }
    case 'bottom':
      return {
        borderLeft: borderStyle,
        borderTop: borderStyle,
        boxShadow: boxShadowBottom,
      }
    case 'right':
      return {
        borderLeft: borderStyle,
        borderTop: borderStyle,
        boxShadow: boxShadowRight,
      }
    case 'left':
      return {
        borderRight: borderStyle,
        borderBottom: borderStyle,
        boxShadow: boxShadowLeft,
      }
  }
}

class TooltipArrow extends PureComponent {
  static propTypes = {
    position: PropTypes.shape({
      left: PropTypes.number,
      top: PropTypes.number,
      placement: PropTypes.string,
    }).isRequired,
    arrowStyle: PropTypes.shape({
      border: PropTypes.string,
      boxShadowRight: PropTypes.string,
      boxShadowBottom: PropTypes.string,
      boxShadowLeft: PropTypes.string,
      boxShadowTop: PropTypes.string,
    }),
    snacksStyle: PropTypes.oneOf(['primary', 'secondary', 'dark']),
  }

  get calculatedStyles() {
    const { position, arrowStyle, snacksStyle } = this.props
    const borderStyle = resolveStylePlacementBorders(snacksStyle, arrowStyle, position.placement)

    return {
      ...styles.arrow,
      ...RESOLVED_STYLE[snacksStyle],
      ...RESOLVED_PLACEMENT[position.placement],
      ...borderStyle,
      left: position.left,
      top: position.top,
    }
  }

  render() {
    return <div style={this.calculatedStyles} />
  }
}

export default TooltipArrow
