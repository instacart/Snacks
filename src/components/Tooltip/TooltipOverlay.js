import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import Portal from '../Portal/Portal'
import TooltipPosition from './TooltipPosition'
import TooltipRootClose from './TooltipRootClose'

class TooltipOverlay extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
    show: PropTypes.bool,
    target: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
    placement: PropTypes.oneOf(['top', 'left', 'right', 'bottom']).isRequired,
    onRootClose: PropTypes.func,
    rootCloseEnabled: PropTypes.bool,
  }

  static defaultProps = {
    rootCloseEnabled: true,
  }

  render() {
    const { show, children, target, placement, onRootClose, rootCloseEnabled } = this.props

    if (!show) {
      return false
    }

    let child = (
      <TooltipPosition target={target} placement={placement}>
        {children}
      </TooltipPosition>
    )

    if (rootCloseEnabled) {
      child = <TooltipRootClose onRootClose={onRootClose}>{child}</TooltipRootClose>
    }

    return <Portal>{child}</Portal>
  }
}

export default TooltipOverlay
