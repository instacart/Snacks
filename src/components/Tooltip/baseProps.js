import PropTypes from 'prop-types'

export default {
  size: PropTypes.oneOf([
    'small',
    'medium',
    'large',
  ]),
  position: PropTypes.oneOf([
    'topLeft',
    'top',
    'topRight',
    'left',
    'right',
    'bottomLeft',
    'bottom',
    'bottomRight',
  ]),
}
