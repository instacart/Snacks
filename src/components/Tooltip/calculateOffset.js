// const OFFSET_TOP = {
//   bottomLeft: 25,
//   bottomRight: 25,
//   bottom: 25,
// }

// const OFFSET_LEFT = {
//   bottomLeft: {
//     small: 0,
//     medium: -180,
//     large: 0,
//   },
//   bottomRight: 25,
//   bottom: 25,
// }

// const OFFSET = {
//   bottomLeft: {
//     top: 25,
//     left: -180
//   }
// }

const calculateTop = ({ postion, size }) => {
  return 0
}

const calculateLeft = ({ postion, size }) => {
  return 0
}


const calculateOffset = ({ postion, size, targetBounds }) => {
  const { top = 0, left = 0 } = targetBounds

  return {
    top: top + calculateTop({ postion, size }),
    left: left + calculateLeft({ postion, size })
  }
}

export default calculateOffset
