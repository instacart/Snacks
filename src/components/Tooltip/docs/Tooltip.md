Example:

      <Tooltip
        target={<Icon name='info' />}
        position='bottomRight'
        size='small'
      >
        Hello
      </Tooltip>
