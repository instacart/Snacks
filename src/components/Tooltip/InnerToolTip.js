import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import Fade from '../Transitions/Fade'
import colors from '../../styles/colors'
import TooltipArrow from './TooltipArrow'
import spacing from '../../styles/spacing'

const styles = {
  root: {
    position: 'relative',
  },
  arrowPadding: {
    top: { paddingBottom: '9px' },
    bottom: { paddingTop: '9px' },
    left: { paddingRight: '9px' },
    right: { paddingLeft: '9px' },
  },
  innerContent: {
    textAlign: 'center',
    borderRadius: 4,
    whiteSpace: 'nowrap',
    fontWeight: 600,
  },
}

const RESOLVED_COLOR = {
  primary: {
    background: colors.GREEN_500,
    color: '#FFF',
    border: `1px solid ${colors.GREEN_500}`,
  },
  secondary: {
    background: '#FFF',
    color: colors.GRAY_46,
    border: `1px solid ${colors.GRAY_74}`,
  },
  dark: {
    background: colors.GRAY_20,
    color: '#FFF',
    border: `1px solid ${colors.GRAY_20}`,
  },
}

const RESOLVED_SIZE = {
  small: {
    fontSize: '14px',
    paddingTop: '9px',
    paddingBottom: '9px',
    ...spacing.PADDING_X_XS,
  },
  medium: {
    fontSize: '16px',
    paddingTop: '9px',
    paddingBottom: '9px',
    ...spacing.PADDING_X_SM,
  },
  large: {
    fontSize: '18px',
    paddingTop: '12px',
    paddingBottom: '12px',
    ...spacing.PADDING_X_MD,
  },
}

class InnerToolTip extends PureComponent {
  static propTypes = {
    children: PropTypes.node,
    snacksStyle: PropTypes.oneOf(['primary', 'secondary', 'dark']),
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    style: PropTypes.shape({
      border: PropTypes.string,
      padding: PropTypes.string,
      boxShadow: PropTypes.string,
    }),
    arrowStyle: PropTypes.shape({
      border: PropTypes.string,
      boxShadowRight: PropTypes.string,
      boxShadowBottom: PropTypes.string,
      boxShadowLeft: PropTypes.string,
      boxShadowTop: PropTypes.string,
    }),
    arrowPosition: PropTypes.shape({}),
  }

  static defaultProps = {
    size: 'medium',
    snacksStyle: 'dark',
  }

  get contentStyles() {
    const { size, style, snacksStyle } = this.props
    return {
      ...styles.innerContent,
      ...RESOLVED_SIZE[size],
      ...RESOLVED_COLOR[snacksStyle],
      ...style,
    }
  }

  renderArrow() {
    const { arrowPosition, arrowStyle, snacksStyle } = this.props
    return (
      <TooltipArrow arrowStyle={arrowStyle} position={arrowPosition} snacksStyle={snacksStyle} />
    )
  }

  render() {
    return (
      <Fade>
        {this.renderArrow()}
        <div style={this.contentStyles}>{this.props.children}</div>
      </Fade>
    )
  }
}

export default InnerToolTip
