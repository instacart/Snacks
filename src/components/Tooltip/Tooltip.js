import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import InnerToolTip from './InnerToolTip'
import TooltipOverlay from './TooltipOverlay'

const noop = () => {} // eslint-disable-line no-empty-function

class Tooltip extends PureComponent {
  static propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    placement: PropTypes.oneOf(['top', 'left', 'right', 'bottom']),
    /**
     * Optional style overrides merged into emotion css
     *
     * @deprecated
     * This prop exists for backwards compatibility and will be
     * removed in a future version
     */
    style: PropTypes.shape({
      border: PropTypes.string,
      padding: PropTypes.string,
      boxShadow: PropTypes.string,
    }),
    arrowStyle: PropTypes.shape({
      border: PropTypes.string,
      boxShadowRight: PropTypes.string,
      boxShadowBottom: PropTypes.string,
      boxShadowLeft: PropTypes.string,
      boxShadowTop: PropTypes.string,
    }),
    target: PropTypes.node.isRequired,
    snacksStyle: PropTypes.oneOf(['primary', 'secondary', 'dark']),
    onDismiss: PropTypes.func,
    onShow: PropTypes.func,
    isVisible: PropTypes.bool,
  }

  static defaultProps = {
    snacksStyle: 'dark',
    placement: 'bottom',
    size: 'small',
    onShow: noop,
    onDismiss: noop,
  }

  state = {
    show: false,
  }

  handleToggle = () => {
    const { onDismiss, onShow } = this.props
    this.setState(
      prevState => ({ show: !prevState.show }),
      () => {
        if (this.state.show) {
          onShow()
        } else {
          onDismiss()
        }
      }
    )
  }

  handleHideToolTip = () => {
    const { onDismiss } = this.props
    this.setState({ show: false })
    onDismiss()
  }

  renderTriggerElement() {
    const { target, isVisible } = this.props
    const { show } = this.state

    if (!target) {
      return
    }
    const extraProps = isVisible == null ? { onClick: this.handleToggle.bind(this) } : {}

    return React.cloneElement(target, {
      ref: node => {
        this.trigger = node
      },
      'aria-haspopup': true,
      'aria-expanded': isVisible || show,
      ...extraProps,
    })
  }

  render() {
    const {
      children,
      className,
      placement,
      size,
      style,
      arrowStyle,
      snacksStyle,
      isVisible,
    } = this.props

    return (
      <div>
        {this.renderTriggerElement()}
        <TooltipOverlay
          placement={placement}
          target={() => this.trigger}
          show={isVisible || this.state.show}
          onRootClose={this.handleHideToolTip}
        >
          <InnerToolTip
            size={size}
            className={className}
            css={style}
            arrowStyle={arrowStyle}
            snacksStyle={snacksStyle}
          >
            {children}
          </InnerToolTip>
        </TooltipOverlay>
      </div>
    )
  }
}

export default Tooltip
