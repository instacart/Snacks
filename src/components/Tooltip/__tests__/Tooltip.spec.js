import React from 'react'
import ReactTestUtils from 'react-dom/test-utils'
import renderer from 'react-test-renderer'
import { StyleRoot } from 'radium'
import { mount } from 'enzyme'
import { spy, stub } from 'sinon'
import Tooltip from '../Tooltip'

describe('Tooltip', () => {
  it('renders Tooltip properly', () => {
    const tree = renderer
      .create(
        <StyleRoot>
          <Tooltip
            target={<div>TRIGGER</div>}
            placement="right"
            size="small"
            snacksStyle="secondary"
          >
            Right Secondary small
          </Tooltip>
        </StyleRoot>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('expect overlay to appear correctly when shown', () => {
    const tooltip = mount(
      <Tooltip
        target={<button>TRIGGER</button>}
        placement="right"
        size="small"
        snacksStyle="secondary"
      >
        Right Secondary small
      </Tooltip>
    )

    const trigger = tooltip.find('button').last()
    trigger.last().simulate('click')

    const tree = renderer.create(tooltip).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should call callbacks', () => {
    const onDismiss = spy()
    const onShow = spy()
    const tooltip = mount(
      <Tooltip
        target={<button>TRIGGER</button>}
        placement="right"
        size="small"
        snacksStyle="secondary"
        onShow={onShow}
        onDismiss={onDismiss}
      >
        Callback tooltip
      </Tooltip>
    )

    const trigger = tooltip.find('button').last()
    trigger.last().simulate('click')
    expect(onShow.calledOnce).toBe(true)
    trigger.last().simulate('click')
    expect(onDismiss.calledOnce).toBe(true)
  })

  it('should not change its internal state if the component is controlled through isVisible', () => {
    const tooltip = mount(<Tooltip target={<button>TRIGGER</button>} isVisible />)

    const trigger = tooltip.find('button').last()
    trigger.last().simulate('click')
    expect(tooltip.props().isVisible).toEqual(true)
    expect(tooltip.state().show).toEqual(false)
  })

  it('should change its internal state when element is uncontrolled', () => {
    const tooltip = mount(<Tooltip target={<button>TRIGGER</button>} />)

    expect(tooltip.state().show).toEqual(false)

    const trigger = tooltip.find('button').last()
    trigger.last().simulate('click')

    expect(tooltip.state().show).toEqual(true)
  })

  it('should not change its internal state when passing in false for isVisible', () => {
    const tooltip = mount(<Tooltip target={<button>TRIGGER</button>} isVisible={false} />)

    const trigger = tooltip.find('button').last()
    trigger.last().simulate('click')

    expect(tooltip.props().isVisible).toEqual(false)
    expect(tooltip.state().show).toEqual(false)
  })

  it('should have false show state and true isVisible prop when not simulating click', () => {
    const tooltip = mount(<Tooltip target={<button>TRIGGER</button>} isVisible />)

    expect(tooltip.props().isVisible).toEqual(true)
    expect(tooltip.state().show).toEqual(false)
  })
})
