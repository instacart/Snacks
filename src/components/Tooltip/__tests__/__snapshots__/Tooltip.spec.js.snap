// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip expect overlay to appear correctly when shown 1`] = `
<div>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    onClick={[Function]}
  >
    TRIGGER
  </button>
</div>
`;

exports[`Tooltip renders Tooltip properly 1`] = `
<div>
  <div
    aria-expanded={false}
    aria-haspopup={true}
    onClick={[Function]}
  >
    TRIGGER
  </div>
</div>
`;
